File: newton_pol.py
def newton_pol(p):
  ip = []
  ip.append(p[0][1])
  for k in range(1, len(p)):
    d = p[k][0] - p[k-1][0]
    u = ip[k - 1]
    for i in range(k - 2, -1, -1):
      u = u * (p[k][0] - p[i][0]) + ip[i]
      d = d * (p[k][0] - p[i][0])
    ip.append((p[k][1] - u) / d)
  return ip

# Newton Polynomial Interpolation
# p_3(x) = 4x^3 + 35x^2 - 84x - 954
# (x, y): (5, 1); (-7, 23); (-6, -54); (0, -954)
# takes in list of points and returns c_k coefficients
print (newton_pol([(5,1), (-7, -23), (-6, -54), (0, -954)]))

File: newton_pol.m
function p = newton_pol(x,y)
  c = zeros(size(x,2), 1);
  c(1) = y(1);
  for k=1:(size(x,2)-1)
      d = x(k+1) - x(k);
      u = c(k);
      for i=(k-2):-1:0
          u = u * (x(k+1) - x(i+1)) + c(i+1);
          d = d * (x(k+1) - x(i+1));
      end
      c(k+1) = (y(k+1) - u) / d;
  end
  p = zeros(1, size(x,2));
  dm = p;
  dm(size(x,2) - 1) = 1;
  dm(size(x,2)) = -x(1);
  p(size(x,2)) = c(1);
  for k=2:size(c)
      if (k > 2)
          dm = conv(dm, [1, -x(k-1)]);
      end
      while(size(dm,2) ~= size(x,2))
          dm(1) = [];
      end
      p = p + c(k) * dm;
  end
end

% INPUT: newton_pol([x1 ... xn], [y1 ... yn])
% OUTPUT: polynomial coefficients using Divided Differences (Newton's method for polynomials)
